### DISCUSSION TITLE
subscriptions / typescript

### DISCUSSION THREAD

**Discussion by jdgamble555** [urql-graphql/urql#1387] [Q&A] [answered]
I am trying to figure out how to do basic subscriptions on typescript or vanilla js from **@urql/core**.

All I see are two options, neither of which return a subscription: **executeSubscription** and **subscription**.

For a query, I do:

```typescript
const subscriptionClient = new SubscriptionClient(`wss://${uri}`, {
  reconnect: true
});
const client = createClient({
  url: `https://${uri}`,
  exchanges: [
    subscriptionExchange({
      forwardSubscription(operation) {
        return subscriptionClient.request(operation);
      },
    }),
  ]
});
// now grab code

const q = client.query(MY_QUERY);
console.log(q);
```
This works, no problem.

For a subscription, I would suspect something like this, to no avail:

```typescript
client.subscription(MY_SUBSCRIPTION_QUERY).subscribe((r: any) => {
  console.log(r);
});
```
But this does not work.

How do I accomplish this?

Thanks,
J

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

Welcome to the urql discussions!

Internally in urql we use `wonka` as a functional stream, this means that when we want to touch these internals we'll have to be aware of that.

```
import { pipe, subscribe } from 'wonka'

const unsub = pipe(
  client.subscription(...)
  subscribe(...callback...)
).unsubscribe
```

As a general note when creating a client, it's advised to use the three default exchanges as well, these are exported by urql.

In docs we talk about subs here https://formidable.com/open-source/urql/docs/advanced/subscriptions/#one-off-subscriptions

**Reply by jdgamble555** [reply]
Perfect.  Thanks @JoviDeCroock !

Is there a way to return the subscription before it is subscribed? That way I could run **rxjs** functions, call the subscription later on my own, etc...

Example:

```typescript
const subscription: Observable<any>;

subscription = pipe(
  client.subscription(GET_TASKS),
  subscribe...
);

subscription.subscribe((result: any) => {
  console.log(result.data);
})
```

**Reply by kitten** [reply]
Wonka is almost exactly like Rx for all intents and purposes here, so the subscription doesn't start until you subscribe. Even the operators are highly similar: https://wonka.kitten.sh/

If you're looking to just drop into Rx then use the `toObservable` operator on a Wonka source. This will give you a spec-compliant JS observable that's interoperable: https://wonka.kitten.sh/api/sinks#toobservable

**Reply by jdgamble555** [reply]
Great.  So last thing.  How do I see ```if error message``` or ```if fetching``` like in the other libraries?

**Reply by kitten** [reply]
That's all represented as one `OperationResult`. So the observables/streams don't have an error state or representation: https://formidable.com/open-source/urql/docs/concepts/errors/

There's no concept of `fetching` in streams. Once you start a stream it is active until you unsubscribe.

**Reply by ollyde** [reply]
@kitten the link was broken, do you have any examples of a connection error?

**Reply by JoviDeCroock** [reply]
This is the working link https://formidable.com/open-source/urql/docs/basics/errors/

**Reply by ollyde** [reply]
@JoviDeCroock thanks! Unfortunately I can't figure out how to get connection errors from Wonka,

```
  // TODO Not sure how to log connection error :-/
const unsub = pipe(
  graphQLClient.subscription(
   // Some sub GQL
  ),
  subscribe(async (message: any) => {
    // Subscribe stuff.
  })
).unsubscribe
```;

**Reply by JoviDeCroock** [reply]
I mean, you won't get them from wonka, it's part of the result of said hook. A subscription and an event-stream are two different things.

When we initiate a subscription we are talking about socket messages, when a message arrives it will go through the exchange-chain (event stream) to your component calling the subscription.

When it fails to connect, the component calling the subscription should get its error property populated

**Reply by ollyde** [reply]
@JoviDeCroock thanks for the clarification, I'm guessing that would be something like ```message.error``` or something then?

**Reply by kitten** [reply]
@ollydixon That's up to the subscription client unfortunately. If there's a problem, it's expected to issue an error event, which would be represented like a network error for the regular `fetchExchange`, meaning that that `OperationResult` you get will have an `error` set to a `CombinedError` from `@urql/core`.
However, afaik not all subscription clients are consistent in letting your message definitively fail since they're mostly not designed for that, and will instead either retry when they regain connectivity or silently fail, unless they have a "passive" callback on the subscription client itself that informs you of a more "global" error, like a complete disconnection.
